sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowUtils.scala

```diff
@@ -71,6 +71,7 @@ object ArrowUtils {
     case d: ArrowType.Decimal => DecimalType(d.getPrecision, d.getScale)
     case date: ArrowType.Date if date.getUnit == DateUnit.DAY => DateType
     case ts: ArrowType.Timestamp if ts.getUnit == TimeUnit.MICROSECOND => TimestampType
+    case date: ArrowType.Date if date.getUnit == DateUnit.MILLISECOND => TimestampType
```

Comment:
Good catch, thanks. Yes, this should be mapped to `Date` in the Arrow schema, not `TimeStamp`.

To give more background, Arrow Dates can have a unit of `DateUnit.DAY` or `DateUnit.MILLISECOND` (see [arrow/vector/types/DateUnit.java#L21-L22](https://github.com/apache/arrow/blob/73d379f4631cd3013371f60876a52615171e6c3b/java/vector/src/main/java/org/apache/arrow/vector/types/DateUnit.java#L21-L22)), currently, if a date with milliseconds is passed, this simply fails; therefore, this change does not affect other type conversions and is fine to map all Arrow dates to Spark dates since now all cases are properly handled.

Reply:
Notice that Spark doesn't have date type with milliseconds, so if we want to map to date type, the hours, minutes, ... will be lost. Otherwise we have to map to timestamp type.
Which is the proper behavior? cc @BryanCutler

Reply:
Right... let me keep just the original PR commit for now. So yes, we could map to date but we would loose time, so the best mapping we have is to `timestamp`.

Let me check back in the `arrow` project as to why `POSIXct` is being mapped to `Date` not `TimeStamp` (see [arrow/r/src/array.cpp#L461](https://github.com/apache/arrow/blob/3a1dd3feb9ca09f92168f46fcfc06c01305df3ec/r/src/array.cpp#L461)), if we can change that then I would agree we don't need this change.

Reply:
I think it should map to `Date` and any extra time would be truncated. Looking at the Arrow format from https://github.com/apache/arrow/blob/master/format/Schema.fbs
```
/// Date is either a 32-bit or 64-bit type representing elapsed time since UNIX
/// epoch (1970-01-01), stored in either of two units:
///
/// * Milliseconds (64 bits) indicating UNIX time elapsed since the epoch (no
///   leap seconds), where the values are evenly divisible by 86400000
/// * Days (32 bits) since the UNIX epoch
```
So it's expected to be a specific number of days without any additional milliseconds.